File: EasyMSXCSharp/EasyMSX/BrokerStrategyParameters.cs
using System.Collections.Generic;
using Name = Bloomberglp.Blpapi.Name;
using LogLevels = com.bloomberg.emsx.samples.Log.LogLevels;
using Request = Bloomberglp.Blpapi.Request;
using Message = Bloomberglp.Blpapi.Message;
using Element = Bloomberglp.Blpapi.Element;
using System.Collections;

namespace com.bloomberg.emsx.samples 
{
    public class BrokerStrategyParameters : IEnumerable<BrokerStrategyParameter>{

	    private static readonly Name 	GET_BROKER_STRATEGY_INFO = new Name("GetBrokerStrategyInfoWithAssetClass");
	    private static readonly Name 	ERROR_INFO = new Name("ErrorInfo");

	    private List<BrokerStrategyParameter> parameters = new List<BrokerStrategyParameter>();
	    internal BrokerStrategy brokerStrategy;
	
	    internal BrokerStrategyParameters(BrokerStrategy brokerStrategy) {
		    this.brokerStrategy = brokerStrategy;
		    loadStrategyParameters();
	    }
	
	    private void loadStrategyParameters() {
		    Log.LogMessage(LogLevels.DETAILED, "Broker Strategy [" + brokerStrategy.parent.broker.name + "." + brokerStrategy.name + "]: Loading strategy parameters");
		    Request request = brokerStrategy.parent.broker.parent.emsxapi.emsxService.CreateRequest(GET_BROKER_STRATEGY_INFO.ToString());
    	    request.Set("EMSX_BROKER", brokerStrategy.parent.broker.name);
    	    request.Set("EMSX_STRATEGY", brokerStrategy.name);
    	    request.Set("EMSX_ASSET_CLASS", brokerStrategy.parent.broker.assetClass.ToString());
    	    brokerStrategy.parent.broker.parent.emsxapi.submitRequest(request,new BrokerStrategyParametersHandler(this));		
	    }
	
	    class BrokerStrategyParametersHandler : MessageHandler {
		
		    BrokerStrategyParameters brokerStrategyParameters;
		
		    internal BrokerStrategyParametersHandler(BrokerStrategyParameters brokerStrategyParameters) {
			    this.brokerStrategyParameters = brokerStrategyParameters;
		    }

		    public void processMessage(Message message) {
			
			    Log.LogMessage(LogLevels.DETAILED, "Broker Strategy Parameters ["+ brokerStrategyParameters.brokerStrategy.parent.broker.name + "." + brokerStrategyParameters.brokerStrategy.name + "]: processing message");
			
	    	    if(message.MessageType.Equals(ERROR_INFO)) {
	        	    Log.LogMessage(LogLevels.DETAILED, "Broker Strategy Parameters ["+ brokerStrategyParameters.brokerStrategy.parent.broker.name + "." + brokerStrategyParameters.brokerStrategy.name + "]: processing RESPONSE error");
	    		    int errorCode = message.GetElementAsInt32("ERROR_CODE");
	    		    string errorMessage = message.GetElementAsString("ERROR_MESSAGE");
	    		    Log.LogMessage(LogLevels.DETAILED, "Broker Strategy Parameters ["+ brokerStrategyParameters.brokerStrategy.parent.broker.name + "." + brokerStrategyParameters.brokerStrategy.name + "]: [" + errorCode + "] " + errorMessage);
	    	    } else if(message.MessageType.Equals(GET_BROKER_STRATEGY_INFO)) {
	        	    Log.LogMessage(LogLevels.DETAILED, "Broker Strategy Parameters ["+ brokerStrategyParameters.brokerStrategy.parent.broker.name + "." + brokerStrategyParameters.brokerStrategy.name + "]: processing succesful RESPONSE");
	    		
	    		    Element parameters = message.GetElement("EMSX_STRATEGY_INFO");
	    		
				    int numValues = parameters.NumValues;
	    		
	    		    for(int i = 0; i < numValues; i++) {
	    			
	    			    Element parameter = parameters.GetValueAsElement(i);

	    			    string parameterName = parameter.GetElementAsString("FieldName");
	    			    int disable = parameter.GetElementAsInt32("Disable");
	    			    string stringValue = parameter.GetElementAsString("StringValue");

	    			    BrokerStrategyParameter newParameter = new BrokerStrategyParameter(brokerStrategyParameters, parameterName,stringValue,disable);
	    			    brokerStrategyParameters.add(newParameter);
	            	    Log.LogMessage(LogLevels.DETAILED, "Broker Strategy Parameters ["+ brokerStrategyParameters.brokerStrategy.parent.broker.name + "." + brokerStrategyParameters.brokerStrategy.name + "] Added new parameter " + parameterName);
	    		    }
	    	    }		
	        }
	    }

        public IEnumerator<BrokerStrategyParameter> GetEnumerator()
        {
            return parameters.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

	    public BrokerStrategyParameter get(int index) {
		    return parameters[index];
	    }
	
	    public void add(BrokerStrategyParameter newBrokerStrategyParameter) {
		    parameters.Add(newBrokerStrategyParameter);
	    }
    }
}

File: Python/easymsx-1.0.0/easymsx/brokerstrategyparameters.py
# brokerstrategyparameters.py

import blpapi
from .brokerstrategyparameter import BrokerStrategyParameter

GET_BROKER_STRATEGY_INFO = blpapi.Name("GetBrokerStrategyInfoWithAssetClass")
ERROR_INFO = blpapi.Name("ErrorInfo")

class BrokerStrategyParameters:
    
    def __init__(self,broker_strategy):
        self.broker_strategy = broker_strategy
        self.parameters=[]
        self.load_strategy_parameters()

    def __iter__(self):
        return self.parameters.__iter__()
    
    def load_strategy_parameters(self):
        
        request = self.broker_strategy.parent.broker.parent.easymsx.emsx_service.createRequest(str(GET_BROKER_STRATEGY_INFO))
        request.set("EMSX_BROKER", self.broker_strategy.parent.broker.name)
        request.set("EMSX_STRATEGY", self.broker_strategy.name)
        request.set("EMSX_ASSET_CLASS",self.broker_strategy.parent.broker.asset_class)
        #print("Sending request: " + str(request))
        
        self.broker_strategy.parent.broker.parent.easymsx.submit_request(request, self.process_message)

    def process_message(self, msg):
        
        if msg.messageType() == ERROR_INFO:
            error_code = msg.getElementAsInteger("ERROR_CODE")
            error_message = msg.getElementAsString("ERROR_MESSAGE")
            if ":2" not in error_message: # fix for issue with Strategies having 0 parameters
                print("GetBrokerStrategyInfoWithAssetClass >> ERROR CODE: %d\tERROR MESSAGE: %s" % (error_code,error_message))

        elif msg.messageType() == GET_BROKER_STRATEGY_INFO:
            
            params = msg.getElement("EMSX_STRATEGY_INFO")
            
            for p in params.values():
                field_name = p.getElementAsString("FieldName")
                disable = p.getElementAsInteger("Disable")
                value = p.getElementAsString("StringValue")
                new_bkr_stp = BrokerStrategyParameter(self, field_name, value, disable)
                self.parameters.append(new_bkr_stp)
                                  
__copyright__ = """
Copyright 2017. Bloomberg Finance L.P.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to
deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:  The above
copyright notice and this permission notice shall be included in all copies
or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.
"""
